libraries/AP_UAVCAN/AP_UAVCAN.h

```diff
@@ -32,253 +37,151 @@
 
-    // Return uavcan from @iface or nullptr if it's not ready or doesn't exist
-    static AP_UAVCAN *get_uavcan(uint8_t iface);
-
-    // this function will register the listening class on a first free channel or on the specified channel
-    // if preferred_channel = 0 then free channel will be searched for
-    // if preferred_channel > 0 then listener will be added to specific channel
-    // return value is the number of assigned channel or 0 if fault
-    // channel numbering starts from 1
-    uint8_t register_gps_listener(AP_GPS_Backend* new_listener, uint8_t preferred_channel);
+    // Return uavcan from @driver_num or nullptr if it's not ready or doesn't exist
+    static AP_UAVCAN *get_uavcan(uint8_t driver_num);
 
-    uint8_t register_gps_listener_to_node(AP_GPS_Backend* new_listener, uint8_t node);
+    void init(uint8_t driver_num) override;
 
-    uint8_t find_gps_without_listener(void);
 
-    // Removes specified listener from all nodes
-    void remove_gps_listener(AP_GPS_Backend* rem_listener);
-
-    // Returns pointer to GPS state connected with specified node.
-    // If node is not found and there are free space, locate a new one
-    AP_GPS::GPS_State *find_gps_node(uint8_t node);
-
-    // Updates all listeners of specified node
-    void update_gps_state(uint8_t node);
-
-    struct Baro_Info {
-        float pressure;
-        float pressure_variance;
-        float temperature;
-        float temperature_variance;
-    };
+    ///// SRV output /////
+    void SRV_push_servos(void);
 
-    uint8_t register_baro_listener(AP_Baro_Backend* new_listener, uint8_t preferred_channel);
-    uint8_t register_baro_listener_to_node(AP_Baro_Backend* new_listener, uint8_t node);
-    void remove_baro_listener(AP_Baro_Backend* rem_listener);
-    Baro_Info *find_baro_node(uint8_t node);
-    uint8_t find_smallest_free_baro_node();
-    void update_baro_state(uint8_t node);
+    ///// LED /////
+    bool led_write(uint8_t led_index, uint8_t red, uint8_t green, uint8_t blue);
 
-    struct Mag_Info {
-        Vector3f mag_vector;
-    };
+    uavcan::Node<0>* get_node() { return _node; }
+    uint8_t get_driver_num() { return _driver_num; }
 
-    uint8_t register_mag_listener(AP_Compass_Backend* new_listener, uint8_t preferred_channel);
-    void remove_mag_listener(AP_Compass_Backend* rem_listener);
-    Mag_Info *find_mag_node(uint8_t node, uint8_t sensor_id);
-    uint8_t find_smallest_free_mag_node();
-    uint8_t register_mag_listener_to_node(AP_Compass_Backend* new_listener, uint8_t node);
-    void update_mag_state(uint8_t node, uint8_t sensor_id);
-
-    struct BatteryInfo_Info {
-        float temperature;
-        float voltage;
-        float current;
-        float remaining_capacity_wh;
-        float full_charge_capacity_wh;
-        uint8_t status_flags;
+    template <typename DataType_>
+    class RegistryBinder {
+    protected:
+        typedef void* (*Registry)(AP_UAVCAN*, uint8_t, const RegistryBinder&);
```

Comment:
I can't that would mean I will have to have a function definition that has message type included in header file. This was done to avoid that, so we can move message related headers to source file, saving a lot of flash space. This is more like a workaround which looks reasonably ok :)

Reply:
`typedef void* (*Registry)(AP_UAVCAN* _ap_uavcan, uint8_t _node_id, const uavcan::ReceivedDataStructure<DataType_>& _msg);`

Why would this need to include the header? It's passed in the operator of the registry binder, so it's basically the same.

Reply:
@OXINARF The answer was to the second part of the query 
Why is a reference to the message (instead of the reference to RegistryBinder) not being passed.

Reply:
It could be forward-declared just like the Cb classes are 😉 But OK, dropping this request.

Reply:
@OXINARF I tried that first, ended up becoming even more messy.